pkg/cmd/util/net.go

```diff
@@ -141,3 +144,53 @@ func TransportFor(ca string, certFile string, keyFile string) (http.RoundTripper
 
 	return &transport, nil
 }
+
+// GetCertificateFunc returns a function that can be used in tls.Config#GetCertificate
+// Returns nil if len(certs) == 0
+func GetCertificateFunc(certs map[string]*tls.Certificate) func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	if len(certs) == 0 {
+		return nil
+	}
+	// Replica of tls.Config#getCertificate logic
+	return func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		if clientHello == nil {
+			return nil, nil
+		}
+
+		name := clientHello.ServerName
+		name = strings.ToLower(name)
+		name = strings.TrimRight(name, ".")
+		for _, candidate := range HostnameMatchSpecCandidates(name) {
+			if cert, ok := certs[candidate]; ok {
+				return cert, nil
+			}
+		}
+		return nil, nil
+	}
+}
+
+// HostnameMatchSpecCandidates returns a list of match specs that would match the provided hostname
+// Returns nil if len(hostname) == 0
+func HostnameMatchSpecCandidates(hostname string) []string {
+	if len(hostname) == 0 {
+		return nil
+	}
+
+	// Exact match has priority
+	candidates := []string{hostname}
+
+	// replace successive labels in the name with wildcards
+	// require an exact match on number of path segments, because certificates cannot wildcard multiple levels of subdomains
+	// this means that a.b.example.com matches *.*.example.com, but not *.example.com
```

Comment:
Give an example where this isn't dumb.  The *.foo.example.com and *.bar.example.com wasn't great, but its something.